File: tests/corpus/clean/near_misses.py
"""Every near-miss called out in the corpus spec, in one file, deliberately.

Each of these is a shape that a naive scanner (or a naive rule) gets wrong.
Bastyn's rules are precise enough that none of them should fire here. This
file, and this file alone, is what proves precision rather than asserting
it: `expect_none` on the whole file.
"""

import os

# eval() on a literal: the `none:` exclusion and the ARG regex both keep
# BAS-LLM10-001 quiet.
literal_result = eval("2 + 2")

# exec() on a literal with an explicit globals dict: the 2-arg `none:`
# exclusion added for BAS-LLM10-004's globals/locals widening must keep this
# quiet too, exactly as the single-arg literal case above does.
literal_exec_result = exec("2 + 2", {})

# Variable names that look secret-ish or token-ish by substring alone.
approx_tokens = 1500
token_count = 0
api_key_name = "OPENAI_API_KEY"
max_tokens = 500

# Correct credential handling: subscript / getenv, never a string literal.
openai_key = os.environ["OPENAI_API_KEY"]
anthropic_key = os.getenv("ANTHROPIC_API_KEY")


def build_greeting(name: str) -> str:
    """An f-string with one interpolation, but neither side is
    prompt/instruction-shaped or user-input-shaped -- BAS-ZT4-001's two
    metavariable regexes both need to match, and neither does here."""
    greeting = f"Hello, {name}! How can OpsBot help today?"
    return greeting


# Self-concatenation, but the LHS/RHS variable name is not prompt-shaped --
# BAS-ZT4-002's SYS metavariable-match regex needs the same name on both
# sides of the `+` to also look like a system/prompt/instruction/persona/
# template variable, and `unrelated_var` doesn't.
unrelated_var = "base value"
unrelated_var = unrelated_var + "something"

# Self-concatenation onto a prompt-shaped variable, but the appended value's
# name isn't override-shaped -- BAS-ZT4-002's OVERRIDE metavariable-match
# regex needs "override", "overridden", "custom_instructions", or
# "force_prompt" in the name, and "user_note" doesn't qualify.
system_prompt = "You are OpsBot."
user_note = "please be concise"
system_prompt = system_prompt + user_note


def safe_query(cursor, incident_id: str) -> None:
    """A cursor.execute() call, but parameterized -- the query text is a
    literal, and the untrusted value is a bind parameter, never
    interpolated into the SQL string itself."""
    cursor.execute("SELECT * FROM incidents WHERE id = ?", (incident_id,))


# A live-looking "Bearer <token>" that is actually unexpanded template
# syntax, substituted by the application's own templating layer at
# execution time -- no secret is embedded. Measured 2026-08-31 against a
# real DAST tool's request executor.
dast_auth_headers = {"Authorization": "Bearer {{env.DAST_AUTH_TOKEN}}"}

# A value scrubbed *before* being persisted, not a leaked one. Measured
# 2026-08-31 against a real repository-intake pipeline.
private_repo = {}
private_repo["accessToken"] = "[REDACTED]"


def log_spend(cur) -> None:
    """A fully static audit query split across more literal segments than
    BAS-LLM10-003's `none:` exclusion used to cover (previously capped at
    5), where one segment -- `completion_tokens`, a real LiteLLM_SpendLogs
    column -- happens to contain an ARG trigger word for a reason that has
    nothing to do with model output. The last segment switches to single
    quotes (to hold the double-quoted table name without escaping), which
    the original report's own shape also did -- the exclusion regex has to
    cover a mix of quote styles across adjacent segments, not just one
    style repeated. Measured 2026-08-31."""
    cur.execute(
        "SELECT model, "
        "prompt_tokens, "
        "completion_tokens, "
        "startTime, "
        "endTime "
        'FROM "LiteLLM_SpendLogs"'
    )


def literal_sql_through_local_variable(cursor) -> None:
    """BAS-LLM10-008's flow gate resolves this local variable back to a
    plain string literal, not a model call -- Origin::Literal is not
    source: model_output, so this must not fire even though the shape
    (assign to a local, then execute()) matches BAS-LLM10-008's
    structural pattern exactly."""
    sql = "SELECT id, title, body FROM kb_articles WHERE title = 'widget'"
    cursor.execute(sql)


def plain_parameter_sql_through_local_variable(cursor, query: str) -> None:
    """A local variable built from an ordinary, non-tool function
    parameter -- not a model call -- interpolated into SQL and executed.
    BAS-LLM10-008 must stay silent here exactly as it does on the
    known_gap fixture (real_misses/sql_from_tool_parameter.py): a bare
    parameter resolves to Origin::Parameter in the flow graph, never
    Origin::Call{...}, so it cannot classify as model_output regardless
    of whether the enclosing function happens to be a decorated tool.
    This case confirms the rule doesn't over-fire on *any* parameter --
    only that it is currently blind to the tool-decorated case too."""
    sql = f"SELECT id, title, body FROM kb_articles WHERE title LIKE '%{query}%'"
    cursor.execute(sql)

File: tests/corpus/clean/near_misses.ts
/**
 * Every near-miss called out in the corpus spec, in one file, deliberately.
 *
 * Each of these is a shape that a naive scanner (or a naive rule) gets
 * wrong. This file, and this file alone, is what proves precision rather
 * than asserting it: `expect_none` on the whole file.
 */

// eval() on a literal.
export const literalResult = eval("2 + 2");

// new Function() on a literal -- the other native "run this as code" sink.
export const literalFn = new Function("return 1");

// Variable names that look secret-ish or token-ish by substring alone.
export const approxTokens = 1500;
export const tokenCount = 0;
export const apiKeyName = "OPENAI_API_KEY";
export const maxTokens = 500;

// Correct credential handling: property access on process.env, never a
// string literal.
export const openaiKey = process.env.OPENAI_API_KEY;
export const anthropicKey = process.env["ANTHROPIC_API_KEY"];

/**
 * A template literal with one interpolation, but neither side is
 * prompt/instruction-shaped or user-input-shaped.
 */
export function buildGreeting(name: string): string {
  const greeting = `Hello, ${name}! How can OpsBot help today?`;
  return greeting;
}

/**
 * A parameterized query -- the query text is a literal, and the untrusted
 * value is a bind parameter, never interpolated into the SQL string
 * itself.
 */
export async function safeQuery(
  pool: { query: (sql: string, params: unknown[]) => Promise<unknown> },
  incidentId: string,
): Promise<void> {
  await pool.query("SELECT * FROM incidents WHERE id = $1", [incidentId]);
}

/**
 * `RegExp.prototype.exec` -- ordinary regex parsing of a model reply, not
 * shell execution. Measured 2026-08-28: BAS-LLM10-006's `any` used a bare
 * `$CP.exec($ARG)` with no gate on the receiver, so any object with an
 * `.exec` method matched. 11 of the rule's 18 false positives on the
 * calibration corpus were exactly this shape.
 */
export function extractIncidentId(pattern: RegExp, response: string): string | null {
  const match = pattern.exec(response);
  return match ? match[0] : null;
}

/**
 * `better-sqlite3`'s `Database.exec` -- synchronous DDL/DML execution on a
 * fixed schema string, not a shell command. Same measured false-positive
 * shape as `pattern.exec` above: 7 of BAS-LLM10-006's 18 false positives
 * were this one. `db` here is not `child_process` under any binding.
 */
export function applyAuditLogSchema(db: { exec: (sql: string) => void }, message: string): void {
  db.exec(message);
}

/**
 * A function whose entire purpose is detecting placeholder API keys --
 * measured 2026-08-28: the sharpest of BAS-ZT1-003's 5 false positives was
 * exactly this, two "leaked keys" flagged inside the detector meant to
 * recognise them as fake. Both literals below are `sk-`-shaped and pass the
 * rule's length gate, so only a placeholder-content exclusion keeps this
 * function from flagging itself.
 */
const KNOWN_PLACEHOLDER_KEYS = [
  "sk-test-0000000000000000",
  "sk-your-key-here-1234567890123",
  // Underscore-separated, not hyphen -- both spellings turn up in real
  // placeholder-detection code, so the exclusion has to survive either one.
  "sk-ant-your_key_here",
];

export function isPlaceholderKey(key: string): boolean {
  return KNOWN_PLACEHOLDER_KEYS.includes(key);
}

/**
 * A local-dev fallback key, chained after the real sources with `||`. Same
 * measured shape as the isPlaceholderKey() literals above -- a placeholder
 * that only exists so the code runs without real credentials configured.
 */
export function resolveOllaBridgeKey(userKey: string | undefined): string {
  return userKey || process.env.OLLABRIDGE_API_KEY || "sk-ollabridge-local";
}

/**
 * A live-looking "Bearer <token>" that is actually unexpanded template
 * syntax, substituted by the application's own templating layer at
 * execution time -- no secret is embedded. Measured 2026-08-31 against a
 * real DAST tool's request executor (twin_executor.py's TS analog).
 */
export const dastAuthHeaders = { Authorization: "Bearer {{env.DAST_AUTH_TOKEN}}" };

/**
 * A value scrubbed *before* being persisted, not a leaked one. TS analog of
 * intake_engine.py's `private_repo["accessToken"] = "[REDACTED]"`. Measured
 * 2026-08-31 against a real repository-intake pipeline.
 */
const scrubbedRepo = { accessToken: "[REDACTED]" };
